System/FilePath/Glob/Match.hs

```diff
@@ -122,14 +124,17 @@ match' o (OpenRange lo hi :xs) path =
 match' o again@(AnyNonPathSeparator:xs) path@(c:cs) =
    match' o xs path || (not (isPathSeparator c) && match' o again cs)
 
-match' o again@(AnyDirectory:xs) path =
-   let parts   = pathParts (dropWhile isPathSeparator path)
-       matches = any (match' o xs) parts || any (match' o again) (tail parts)
-    in if null xs && not (matchDotsImplicitly o)
-          --  **/ shouldn't match foo/.bar, so check that remaining bits don't
-          -- start with .
-          then all (not.isExtSeparator.head) (init parts) && matches
-          else matches
+match' o (AnyDirectory:xs) path =
+   if not (matchDotsImplicitly o)
+      --  **/ shouldn't match foo/.bar, so check that remaining bits don't
+      -- start with .
+      then all (not.isExtSeparator.head) matchedDirs && hasMatch
+      else hasMatch
+ where parts   = pathParts (dropWhile isPathSeparator path)
+       matches = map (match' o xs) parts
+       hasMatch = or matches
+       matchIndex = fromMaybe 0 (elemIndex True matches)
```

Comment:
As a tiny performance thing, I think referencing `matches` twice like this is likely to prevent the intermediate list from getting optimized away. I'm not entirely sure though, feel free to examine the compiler output if you want to check. An alternative would be:

```haskell
matchIndex = elemIndex True matches
hasMatch = isJust matchIndex
(matchedDirs, _) = splitAt (fromMaybe 0 matchIndex) parts
```

At that point you could also go about things more directly and get rid of `matches`, using `findIndex` (or whatever it's called) instead of `elemIndex` on top of `map`.